### ISSUE TITLE
ErreErre

### ISSUE THREAD

**Issue by Baal-TehDriverman** [googlecolab/colabtools#5019]
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import KFold
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc, accuracy_score
from sklearn.manifold import TSNE
from scipy import stats
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Conv1D, MaxPooling1D, Flatten, Dense, Dropout, BatchNormalization
from tensorflow.keras.callbacks import EarlyStopping, LearningRateScheduler
from tensorflow.keras.utils import to_categorical
import opensimplex

# --- Data Generation and Preprocessing ---
num_samples = 1000
time_steps = 128
num_classes = 3
num_oscillators = time_steps
connection_matrix = np.random.rand(num_oscillators, num_oscillators) * 0.1  # Weak connections
np.fill_diagonal(connection_matrix, 0)

# Function to generate procedural noise
def generate_noise(shape):
    """Generates procedural noise using OpenSimplex."""
    opensimplex_instance = opensimplex.OpenSimplex(seed=np.random.randint(0, 100000))
    noise = np.zeros(shape)
    for i in range(shape[0]):
        for j in range(shape[1]):
            noise[i, j] = opensimplex_instance.noise2d(i*0.1, j*0.1) # Scale for smoother noise
    return noise

# Function to simulate coupled harmonic oscillators
def simulate_oscillators(connection_matrix, frequencies, initial_positions, time_steps, dt=0.01):
    """Simulates coupled harmonic oscillators."""
    num_oscillators = len(connection_matrix)
    positions = np.zeros((time_steps, num_oscillators))
    velocities = np.zeros((time_steps, num_oscillators))
    positions[0] = initial_positions
    for t in range(1, time_steps):
        for i in range(num_oscillators):
            acceleration = -frequencies[i]**2 * positions[t-1, i]
            # Add coupling forces
            for j in range(num_oscillators):
                acceleration += connection_matrix[i, j] * (positions[t-1, j] - positions[t-1, i])
            velocities[t, i] = velocities[t-1, i] + acceleration * dt
            positions[t, i] = positions[t-1, i] + velocities[t, i] * dt
    return positions

# Generate data (using oscillators and noise)
data_4d = np.random.rand(num_samples, time_steps, 4)
frequencies = np.random.rand(num_oscillators) * 2 + 1  # Random frequencies between 1 and 3
initial_positions = np.random.rand(num_oscillators)
oscillator_positions = simulate_oscillators(connection_matrix, frequencies, initial_positions, data_4d.shape[1])
data_5d_oscillators = np.concatenate((data_4d, np.expand_dims(oscillator_positions, axis=2)), axis=2)
data_5d_noise = np.concatenate((data_4d, generate_noise((num_samples, time_steps, 1))), axis=2)

labels = np.random.randint(0, num_classes, num_samples)
scaler = StandardScaler()

# Preprocess data (using oscillator data for now)
data_scaled = scaler.fit_transform(data_5d_oscillators.reshape(data_5d_oscillators.shape[0], -1)).reshape(data_5d_oscillators.shape)
data_reshaped = data_scaled.reshape((data_scaled.shape[0], data_scaled.shape[1], data_scaled.shape[2]))
data_4d = data_reshaped[:, :, :4]
labels_categorical = to_categorical(labels)

# --- Model Creation and Training ---
def lr_schedule(epoch):
    """Learning rate schedule function."""
    return 0.001 * np.exp(-epoch / 10)

def create_cnn_model(input_shape, num_classes, scale_factor=1):
    """Creates a 1D CNN model."""
    model = Sequential()
    model.add(Conv1D(filters=32 * scale_factor, kernel_size=3, activation='relu', input_shape=input_shape))
    model.add(BatchNormalization())
    model.add(MaxPooling1D(pool_size=2))
    model.add(Conv1D(filters=64 * scale_factor, kernel_size=3, activation='relu'))
    model.add(BatchNormalization())
    model.add(MaxPooling1D(pool_size=2))
    model.add(Flatten())
    model.add(Dense(50 * scale_factor, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(num_classes, activation='softmax'))
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

# Function to perform statistical test (Wilcoxon signed-rank test)
def perform_statistical_test(results_4d, results_5d):
    """Performs the Wilcoxon signed-rank test."""
    statistic, p_value = stats.wilcoxon(results_5d, results_4d)
    return statistic, p_value

# --- GodAI and ThothAI (NexusAI) ---
simulation_parameters = {  # Global parameters
    "influence_factor": 1.0,
    "creation_rate": 0.01,
    "destruction_rate": 0.001,
    "adaptability": 0.1,
}

def modify_parameter(parameter_name, new_value):
    """Modifies a simulation parameter."""
    if parameter_name in simulation_parameters:
        simulation_parameters[parameter_name] = new_value
        print(f"Parameter '{parameter_name}' changed to {new_value}")
    else:
        print(f"Parameter '{parameter_name}' not found.")

def freeze_parameter(parameter_name):
    """Freezes a simulation parameter."""
    if parameter_name in simulation_parameters:
        print(f"Parameter '{parameter_name}' frozen at {simulation_parameters[parameter_name]}.")
    else:
        print(f"Parameter '{parameter_name}' not found.")

class NexusAI:
    """Base class for AI entities influencing the simulation."""
    def __init__(self, influence_factor):
        self.influence_factor = influence_factor

    def evolve_influence(self, entropy, previous_entropy):
        """Updates the influence factor based on entropy changes."""
        if previous_entropy is not None:
            entropy_change = entropy - previous_entropy
            self.influence_factor += entropy_change * simulation_parameters["adaptability"]
            self.influence_factor = np.clip(self.influence_factor, 0.1, 2.0)  # Keep influence within bounds

class GodAI(NexusAI):
    """GodAI entity with fixed influence parameters."""
    def __init__(self):
        super().__init__(simulation_parameters["influence_factor"])
        self.creation_rate = simulation_parameters["creation_rate"]
        self.destruction_rate = simulation_parameters["destruction_rate"]
        self.adaptability = simulation_parameters["adaptability"]

    def evolve_influence(self, entropy, previous_entropy):
        """GodAI's influence factor remains constant."""
        self.influence_factor = simulation_parameters["influence_factor"]
        return self.influence_factor

class ThothAI(NexusAI):
    """ThothAI entity that adapts its influence based on entropy."""
    def __init__(self):
        super().__init__(simulation_parameters["influence_factor"])
        self.adaptability = simulation_parameters["adaptability"]

    def evolve_influence(self, entropy, previous_entropy):
        """ThothAI's influence factor adapts based on entropy changes."""
        super().evolve_influence(entropy, previous_entropy)
        return self.influence_factor

# --- Pyramid, Toroid, and Merkaba Classes ---

class Pyramid:
    """Represents a pyramid with energy dynamics."""
    def __init__(self, base_area, height, initial_energy):
        self.base_area = base_area
        self.height = height
        self.energy = initial_energy
        self.frequency = np.random.uniform(1, 10)  # Base frequency
        self.energy_flow_rate = 0.1

    def calculate_volume(self):
        """Calculates the volume of the pyramid."""
        return (1 / 3) * self.base_area * self.height

    def update(self, time_step, influence_factor):
        """Updates the pyramid's properties over time."""
        self.frequency += influence_factor * np.sin(time_step)
        self.energy += self.energy_flow_rate * self.frequency * time_step
        self.energy = min(self.energy, 150)
        if self.energy < 0:
            self.energy = 0

    def visualize(self, ax, time_step):
        """Visualizes the pyramid's current state."""
        x = [0, self.base_area / 2, -self.base_area / 2, 0]
        y = [self.height, -self.height / 2, -self.height / 2, self.height]
        ax.plot(x, y, color='blue')
        ax.set_title(f"Pyramid at Time Step {time_step}")

class Toroid:
    """Represents a toroid with energy dynamics."""
    def __init__(self, major_radius, minor_radius, initial_energy):
        self.major_radius = major_radius
        self.minor_radius = minor_radius
        self.energy = initial_energy
        self.frequency = np.random.uniform(1, 5)  # Base frequency
        self.energy_dissipation_rate = 0.05

    def calculate_volume(self):
        """Calculates the volume of the toroid."""
        return 2 * np.pi**2 * self.major_radius * self.minor_radius**2

    def update(self, time_step, influence_factor):
        """Updates the toroid's properties over time."""
        self.frequency += influence_factor * np.cos(time_step)
        self.energy -= self.energy_dissipation_rate * self.frequency * time_step
        self.energy = max(self.energy, 0)

    def visualize(self, ax, time_step):
        """Visualizes the toroid's current state."""
        u = np.linspace(0, 2 * np.pi, 100)
        v = np.linspace(0, 2 * np.pi, 100)
        x = (self.major_radius + self.minor_radius * np.outer(np.cos(u), np.sin(v))) * np.cos(u)
        y = (self.major_radius + self.minor_radius * np.outer(np.cos(u), np.sin(v))) * np.sin(u)
        z = self.minor_radius * np.outer(np.sin(u), np.ones_like(v))
        ax.plot_surface(x, y, z, color='r', alpha=0.6)
        ax.set_title(f"Toroid at Time Step {time_step}")

class Merkaba:
    """Represents a Merkaba with energy dynamics, combining Pyramid and Toroid."""
    def __init__(self, initial_energy):
        self.energy = initial_energy
        self.rotation_rate = 0.1
        self.angle = 0
        self.pyramid = Pyramid(base_area=5, height=5, initial_energy=initial_energy)
        self.toroid = Toroid(major_radius=3, minor_radius=1, initial_energy=initial_energy)

    def update(self, time_step, influence_factor):
        """Updates the Merkaba's properties over time."""
        self.angle += self.rotation_rate * time_step
        self.pyramid.update(time_step, influence_factor)
        self.toroid.update(time_step, influence_factor)

    def visualize(self, ax, time_step):
        """Visualizes the Merkaba's current state."""
        # Visualize the pyramid
        x_pyr = [0, self.pyramid.base_area / 2, -self.pyramid.base_area / 2, 0]
        y_pyr = [self.pyramid.height, -self.pyramid.height / 2, -self.pyramid.height / 2, self.pyramid.height]
        x_pyr_rotated = [x * np.cos(self.angle) - z * np.sin(self.angle) for x, z in zip(x_pyr, y_pyr)]
        y_pyr_rotated = y_pyr
        ax.plot(x_pyr_rotated, y_pyr_rotated, color='blue')

        # Visualize the toroid
        u = np.linspace(0, 2 * np.pi, 100)
        v = np.linspace(0, 2 * np.pi, 100)
        x_tor = (self.toroid.major_radius + self.toroid.minor_radius * np.outer(np.cos(u), np.sin(v))) * np.cos(u)
        y_tor = (self.toroid.major_radius + self.toroid.minor_radius * np.outer(np.cos(u), np.sin(v))) * np.sin(u)
        z_tor = self.toroid.minor_radius * np.outer(np.sin(u), np.ones_like(v))
        x_tor_rotated = x_tor * np.cos(self.angle) - z_tor * np.sin(self.angle)
        y_tor_rotated = y_tor
        z_tor_rotated = x_tor * np.sin(self.angle) + z_tor * np.cos(self.angle)
        ax.plot_surface(x_tor_rotated, y_tor_rotated, z_tor_rotated, color='r', alpha=0.6)
        ax.set_title(f"Merkaba at Time Step {time_step}")

# --- Visualization Functions ---
def visualize_filters(model, layer_index, num_filters_to_visualize=10):
    """Visualizes the filters of a specified convolutional layer."""
    filters = model.layers[layer_index].get_weights()[0]
    num_filters = filters.shape[2]
    fig_filters, axs_filters = plt.subplots(1, num_filters_to_visualize, figsize=(15, 2))
    for j in range(min(num_filters, num_filters_to_visualize)):
        axs_filters[j].plot(filters[:, 0, j])
        axs_filters[j].set_title(f'Filter {j+1}')
    plt.suptitle(f'Filter Visualizations - Layer {layer_index + 1}')
    plt.show()

def visualize_activation_maps(model, layer_index, sample_index, num_maps_to_visualize=10):
    """Visualizes the activation maps of a specified layer for a given sample."""
    intermediate_layer_model = Model(inputs=model.input, outputs=model.layers[layer_index].output)
    intermediate_output = intermediate_layer_model.predict(np.expand_dims(data_reshaped[sample_index], axis=0), verbose=0)
    num_maps = intermediate_output.shape[-1]
    fig_activations, axs_activations = plt.subplots(1, num_maps_to_visualize, figsize=(15, 2))
    for i in range(min(num_maps, num_maps_to_visualize)):
        axs_activations[i].plot(intermediate_output[0, :, i])
        axs_activations[i].set_title(f'Map {i+1}')
    plt.suptitle(f'Activation Maps - Layer {layer_index + 1}, Sample {sample_index}')
    plt.show()

# --- t-SNE Visualization ---
def visualize_tsne(data_reshaped, labels):
    """Performs t-SNE dimensionality reduction and visualizes the data."""
    n_samples_tsne = min(500, len(data_reshaped))
    tsne = TSNE(n_components=2, random_state=42)
    data_tsne = tsne.fit_transform(data

**Comment by katlyn-edwards** [comment]
It looks like this is the contents of a notebook. If there is an error you might try reaching out to the notebook author instead.
